### PR TITLE
Elements: Layouts - Take #2

### PR DESCRIPTION
**Pull request by talbs**
**UPDATE: V2 of https://github.com/edx/ux-pattern-library/pull/117 - Freshly rebased against the Grid PR**

- - -

##### This work, based off of https://github.com/edx/ux-pattern-library/pull/105, adds in a set of basic layouts for use in views, including:

* default stacking/collapsing of columns for our smaller breakpoints
* in-page class-based layouts for bands of content
* options for layout settings, such as flush columns and reverse order
* view-based layouts for larger controls of pages (including full width/bleed layouts (prefixed with ``layout-view``)
* an example view (``/examples/layouts.html``)

- - -

##### What this system/work does
* Sets up layout settings for different uses (on grid with external gutters, flush columns, reversed order)
* Relies on Susy and Breakpoint syntax 
* Offers basic responsive layout support in small viewport cases

- - -

##### What this system/work does not do
* Allow for nesting of layouts - one layout must be ended before another can be started
* Provide DOM-based classes for all possible layouts imaginable (easily grid-divisible and common UI convention cases only)
* Provide nuanced breakpoint cases for specific layout needs
* Creates vertical rhythm between bands of layouts

- - -

##### Reviewers
* [ ] FED - @frrrances 
* [ ] Usage - @marcotuts 

- - - 

FYI, @dsjen, @andy-armstrong , and @clrux.

### CODE REVIEW

**Comment by talbs** [comment]
FYI, this has been rebased against master (which now has the Grid System merged in). Commits and diffs here should now just be layouts-centric.

File: _posts/elements/2015-07-10-layouts.md [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+            <span class="example-label">col-b</span>
+        </main>
+
+        <aside class="layout-col layout-col-c">
+            <span class="example-label">col-c</span>
+        </aside>
+    </div>
+</div>
+
+<h3 class="hd-6 example-set-hd">1/4 + 1/2 + 1/4 Layout (Reversed)</h3>
```
Comment (frrrances) [inline _posts/elements/2015-07-10-layouts.md]: Isn't this the same as regular (non-reversed)?

File: _posts/elements/2015-07-10-layouts.md [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+            <span class="example-label">col-b</span>
+        </main>
+
+        <aside class="layout-col layout-col-c">
+            <span class="example-label">col-c</span>
+        </aside>
+    </div>
+</div>
+
+<h3 class="hd-6 example-set-hd">1/3 + 1/3 + 1/3 Layout (Reversed)</h3>
```
Comment (frrrances) [inline _posts/elements/2015-07-10-layouts.md]: Isn't this the same as regular (non-reversed)?

File: _posts/elements/2015-07-10-layouts.md [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+            <span class="example-label">col-a</span>
+        </main>
+
+        <aside class="layout-col layout-col-b">
+            <span class="example-label">col-b</span>
+        </aside>
+    </div>
+</div>
+
+<h3 class="hd-6 example-set-hd">1/2 + 1/2 Layout (Reversed)</h3>
```
Comment (frrrances) [inline _posts/elements/2015-07-10-layouts.md]: Isn't this the same as regular (non-reversed)?

File: _posts/elements/2015-07-10-layouts.md [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+            <span class="example-label">col-b</span>
+        </main>
+
+        <aside class="layout-col layout-col-c">
+            <span class="example-label">col-c</span>
+        </aside>
+    </div>
+</div>
+
+<h3 class="hd-6 example-set-hd">1/4 + 1/4 + 1/2 Layout Layout (Reversed)</h3>
```
Comment (frrrances) [inline _posts/elements/2015-07-10-layouts.md]: ooo... the elusive layout layout! i spotted it in the wild!

File: _src/pattern-library/sass/components/_layouts.scss [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+
+            .layout-col-b {
+                @include span(9 of 12 last);
+            }
+        }
+    }
+}
+
+// 1/3 + 1/3 + 1/3 Layout
+.layout-1t1t1t {
```
Comment (frrrances) [inline _src/pattern-library/sass/components/_layouts.scss]: nice layout name!

File: _src/pattern-library/sass/components/_layouts.scss [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+
+            .layout-col-c {
+                @include span(6 of 12 last);
+            }
+        }
+    }
+}
+
+// ----------------------------
+// #CANNED - VIEW
```
Comment (frrrances) [inline _src/pattern-library/sass/components/_layouts.scss]: nice! love the view layouts.

**Comment by frrrances** [comment]
@talbs Nice work! This is going to be really awesome to put to use. I noted a few things I thought were repetitive, but if I'm missing something, feel free to ignore. I think we will want to update how the layouts flex for responsive once we have some real world examples, but I think what you've laid down is a good starting point. Couldn't find much to argue with... 

:+1:

File: _posts/elements/2015-07-10-layouts.md (hunk 1 above)
Comment (talbs) [inline _posts/elements/2015-07-10-layouts.md]: Aha! I fooled you on this. There is one difference, this layout has a class of ``.layout-reversed`` which reverses the flow of the layout (without needing to shift any DOM order at all - that's why no other markup has changed). 

I thought this way of reversing a layout was a nice option especially if its just for presentation. Seem beneficial to you or should we nix reverse layouts and just have folks adjust the DOM?

File: _posts/elements/2015-07-10-layouts.md (hunk 2 above)
Comment (talbs) [inline _posts/elements/2015-07-10-layouts.md]: See [my explanation elsewhere about reverse layouts](https://github.com/edx/ux-pattern-library/pull/121#discussion_r35204556).

File: _posts/elements/2015-07-10-layouts.md (hunk 1 above)
Comment (talbs) [inline _posts/elements/2015-07-10-layouts.md]: Sorry, should have made an in-line code note about this technique.

File: _posts/elements/2015-07-10-layouts.md (hunk 3 above)
Comment (talbs) [inline _posts/elements/2015-07-10-layouts.md]: See [my explanation elsewhere about reverse layouts](https://github.com/edx/ux-pattern-library/pull/121#discussion_r35204556).

File: _posts/elements/2015-07-10-layouts.md (hunk 4 above)
Comment (talbs) [inline _posts/elements/2015-07-10-layouts.md]: Sweet cheese and rice! #typotalbs

File: _src/pattern-library/sass/components/_layouts.scss (hunk 5 above)
Comment (talbs) [inline _src/pattern-library/sass/components/_layouts.scss]: Nice, @frrrances. Very mature of you!
Should I rename it to make sure this PR gets a PG-13 ratings?

**Comment by talbs** [comment]
@frrrances, thanks for the review! I've addressed your feedback. Let me know if there's anything else you need during a hopefully quick second pass.

- - -

> I think we will want to update how the layouts flex for responsive once we have some real world examples...

^ Yes, I completely agree. Once we have a header, navigation, or specific utility/per-app views, we can add some more details/cases to the responsive logic. I'll be on the lookout to make those changes.

**Comment by talbs** [comment]
@marcotuts, pinged you originally as a usage reviewer, but I think you're OOO currently. I'll merge this and then address any feedback you have afterwards, if that's alright by you.

**Pull request merged**